### PR TITLE
[NTOS:CC] Restore unlock/reacquire locks around MmPageOutPhysicalAddress

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_Put back remove locks and reacquire locks around MmPageOutPhysicalAdress._

JIRA issue: maybe somewhat related to CORE17624, but reactosfanboy doesn't want the important commits of that to be spammed with it. Therefore only mentioned here in the text.

## Proposed changes

_The original code for CcRosTrimCache had this code._
_The committed PR mistakenly had a patch removing these unlock and reacquire locks_
_This was my mistake._
_This PR puts them back in place._

Revision Note: I added an ASSERT(FALSE) and a DPRINT1 into the main code path here and tested this.
During all of my testing on the items below, I was never able to trigger the ASSERT or get the DPRINT1 message.

1) I ran the large file copy test on 512 K memory as in https://jira.reactos.org/browse/CORE-17624.
2) I ran a full ReactOS regression test suite locally.
3) I did a full compile of ReactOS on ReactOS.

Based on the results of my testing and the apparent code issues here, I have proposed disabling this code.
I kept the original code in the 'else' path for the chance that someone can find a way to trigger this code path.
We will see what the developers familiar with the Memory Manager think about this option. Thanks.

### CODE REVIEW

**Review by ThFabba — APPROVED** [review]

**Comment by JoachimHenze** [comment]
We are not aware yet of any specific or reproducible testcase that would be fixed by this PR.
The random reboots on the testbots will **not** be fixed by that PR yet:
```
VBox: https://reactos.org/testman/compare.php?ids=89685,89689,89691,89692,89694
VBox: https://reactos.org/testman/compare.php?ids=89685,89697,89703,89705,89706,89708

KVM:  https://reactos.org/testman/compare.php?ids=89684,89688,89690,89693,89695
KVM:  https://reactos.org/testman/compare.php?ids=89684,89698,89704,89707,89709

VBox/21612 - Did not complete
VBox/21613 - Extra reboot
VBox/21614 - Extra reboot
VBox/21615 - LGTM
VBox/21616 - LGTM
VBox/21617 - LGTM
VBox/21619 - LGTM
VBox/21620 - LGTM
VBox/21621 - Crash in mshtml:script


KVM/34139 - Extra reboot
KVM/34140 - Did not complete
KVM/34141 - Extra reboot
KVM/34142 - LGTM
KVM/34143 - LGTM
KVM/34144 - LGTM
KVM/34146 - LGTM
KVM/34147 - Did not complete
KVM/34148 - Crash in msi:action (known to happen randomly atm on current master head)
    *** Assertion failed: Refs > 0
    ***   Source File: /srv/buildbot/worker_data/Build_GCCLin_x86/build/ntoskrnl/cc/view.c, line 604

```
Only justification up to now comes from Thomas Fabers thoughts:
`MmPageOutPhysicalAddress cannot be called with a spinlock held, it's against the contract. It will assert if it ever gets to the MmLockAddressSpace line: https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/mm/rmap.c;hb=1b3eed58cac4cd7f8ebfeecbe1f244076c83a8b8#l103 If that's not happening in your testing, then the code just never gets that far inside the function, but it's still wrong`

In case anyone would ever identify a real-world-scenario being fixed by it, please mention it here!

File: ntoskrnl/cc/view.c [hunk 1]
```diff
@@ -497,13 +497,21 @@ CcRosTrimCache(
         /* Check if it's mapped and not dirty */
         if (InterlockedCompareExchange((PLONG)&current->MappedCount, 0, 0) > 0 && !current->Dirty)
         {
+            /* We have to break these locks to call MmPageOutPhysicalAddress */
+            KeReleaseSpinLockFromDpcLevel(&current->SharedCacheMap->CacheMapLock);
+            KeReleaseQueuedSpinLock(LockQueueMasterLock, oldIrql);
```
Comment (tkreuzer) [inline ntoskrnl/cc/view.c:502]: Breaking this lock will make any pointer to VACBs in `VacbLruListHead` potentially invalid. While there is a reference on `current`, preventing it from beig deleted, there is no reference on `current_entry`. This means after breaking the lock, you have to either start over at `VacbLruListHead.Flink`, which can potentially lead to an infinite loop. or take a reference on the next VACB.

**Review by tkreuzer — APPROVED** [review]

**Review by tkreuzer — APPROVED** [review]

**Review by JoachimHenze — APPROVED** [review]
Ok, if Timo says there **is** evidence in this PR improving something (that is how I do understand his reply and approval), then ok also from my side. I at least couldn't identify anything that it would break.
But I did remove the relation to the already resolved CORE-17624, I don't think it would be good/beneficial to have a linking to that.

File: ntoskrnl/cc/view.c [hunk 2]
```diff
@@ -487,7 +485,6 @@ CcRosTrimCache(
         current = CONTAINING_RECORD(current_entry,
                                     ROS_VACB,
                                     VacbLruListEntry);
-        current_entry = current_entry->Flink;
```
Comment (HBelusca) [inline ntoskrnl/cc/view.c:490]: Is this move of the current_entry to the very end of the while-loop really needed?

**Review by HBelusca — COMMENTED** [review]
Is it possible to exercise this code by testing ReactOS on a very-low memory scenario in VM?

File: ntoskrnl/cc/view.c (hunk 2 above)
Comment (Doug-Lyons) [reply ntoskrnl/cc/view.c:490]: I do not really know, but that is what Timo says. I do trust him on this. Thanks.

File: ntoskrnl/cc/view.c (hunk 1 above)
Comment (Doug-Lyons) [reply ntoskrnl/cc/view.c:507]: @HBelusca, Here is where Timo says that this statement should be moved. Thanks.

**Comment by Doug-Lyons** [comment]
> Is it possible to exercise this code by testing ReactOS on a very-low memory scenario in VM?

I will try and do some testing regarding your question, but I have already seen that under low memory conditions, it is still possible to lock up ReactOS in spite of all of our efforts in this area so far.

**Comment by Doug-Lyons** [comment]
Thanks for everyone's help with this.

**Pull request merged**